src/chainstate/stacks/boot/bns.clar

```diff
@@ -0,0 +1,973 @@
+        (err ERR_NAME_EXPIRED))
+      true)    
+    ;; The amount burnt must be equal to or greater than the cost of the namespace
+    (asserts!
+      (>= stx-to-burn (compute-name-price name (get price-function namespace-props)))
+      (err ERR_NAME_STX_BURNT_INSUFFICIENT))
+    ;; The name must not be revoked
+    (asserts!
+      (is-none (get revoked-at name-props))
+      (err ERR_NAME_REVOKED))
+    ;; Transfer the name, if any new-owner
+    (if (is-none new-owner)
+      (ok true) 
+      (let ((owner-unwrapped (unwrap! new-owner (err ERR_PANIC))))
+        (let ((current-owned-name (map-get? owner-name ((owner owner-unwrapped)))))
+          (let ((can-new-owner-get-name (if (is-none current-owned-name)
+                                  true
+                                  (unwrap! 
+                                    (is-name-lease-expired?
+                                      (unwrap! (get namespace current-owned-name) (err ERR_PANIC))
+                                      (unwrap! (get name current-owned-name) (err ERR_PANIC)))
+                                    (err ERR_PANIC)))))
+            (asserts!
+              can-new-owner-get-name
+              (err ERR_PRINCIPAL_ALREADY_ASSOCIATED))
+            (unwrap!
+              (update-name-ownership? namespace name contract-caller owner-unwrapped)
+              (err ERR_NAME_COULD_NOT_BE_TRANSFERED))
+            (ok true)))))
+    ;; Update the zonefile, if any.
+    (if (is-none zonefile-hash)
+      (map-set name-properties
+        ((namespace namespace) (name name))
+        ((registered-at (some block-height))
+         (imported-at none)
+         (revoked-at none)
+         (zonefile-hash (get zonefile-hash name-props))))
+      (update-zonefile-and-props
+              namespace 
+              name
+              (some block-height)
+              none
+              none
+              (unwrap! zonefile-hash (err ERR_PANIC))))  
+    (ok true)))
+
+;; Additionals public methods
+
+(define-read-only (can-name-be-registered (namespace (buff 19)) (name (buff 16)))
+  (let (
+      (wrapped-name-props (map-get? name-properties ((namespace namespace) (name name))))
+      (current-owner (map-get? owner-name ((owner contract-caller))))
+      (namespace-props (unwrap! (map-get? namespaces ((namespace namespace))) (ok false))))
+    ;; Ensure that namespace has been launched 
+    (unwrap! (get launched-at namespace-props) (ok false))
+    ;; Early return - Name has never be minted
+    (asserts! (is-none (nft-get-owner? names (tuple (name name) (namespace namespace)))) (ok true))
+    ;; Integrity check - Ensure that we have some entries in nft && owner-name && name-props
+    (asserts! 
+      (and 
+        (is-some wrapped-name-props)
+        (is-some current-owner))
+      (err ERR_PANIC))
+    (let ((name-props (unwrap! wrapped-name-props (err ERR_PANIC))))
+      ;; Early return - Name has been revoked and can be registered
```

Comment:
I don't think this is right.  In Stacks 1.0, a name can only be registered after it expires, even if it has been revoked.

Reply:
Addressed with https://github.com/blockstack/stacks-blockchain/pull/2013/commits/5241bdf61b21d7f855df709cfd157a89a7147a0b

Reply:
Not quite -- see https://github.com/blockstack/stacks-blockchain/pull/2013/files#r538997373